kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt

```diff
@@ -1,72 +1,1061 @@
+import io.kotlintest.matchers.doubles.shouldBeNegative
+import io.kotlintest.matchers.doubles.shouldBePositive
+import io.kotlintest.matchers.doubles.shouldNotBeBetween
+import io.kotlintest.matchers.doubles.shouldNotBeExactly
+import io.kotlintest.matchers.doubles.shouldNotBeGreaterThan
+import io.kotlintest.matchers.doubles.shouldNotBeGreaterThanOrEqual
+import io.kotlintest.matchers.doubles.shouldNotBeLessThan
+import io.kotlintest.matchers.doubles.shouldNotBeLessThanOrEqual
+import io.kotlintest.matchers.doubles.shouldNotBeNegative
+import io.kotlintest.matchers.doubles.shouldNotBePositive
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.assertAll
+import io.kotlintest.should
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldNot
 import io.kotlintest.shouldNotBe
-import io.kotlintest.specs.ShouldSpec
-import io.kotlintest.tables.*
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FreeSpec
+import kotlin.Double.Companion.MAX_VALUE
+import kotlin.Double.Companion.NEGATIVE_INFINITY
+import kotlin.Double.Companion.NaN
+import kotlin.Double.Companion.POSITIVE_INFINITY
+import kotlin.math.absoluteValue
 
-class DoubleMatchersTest : ShouldSpec() {
-    init {
-
-        should("Not be positive") {
-            val acceptedValues = table(
-                    headers("a"),
-                    row(-1.0),
-                    row(-0.00001),
-                    row(-42.0),
-                    row(0.0),
-                    row(Double.NaN),
-                    row(Double.NEGATIVE_INFINITY)
-            )
-
-            forAll(acceptedValues) {
-                it.shouldNotBePositive()
-                it shouldNotBe positive()
+class DoubleMatchersTest : FreeSpec() {
+  
+  private val nonNumericDoubles = listOf(NaN, POSITIVE_INFINITY, NEGATIVE_INFINITY)
+  
+  private val numericDoubles = Gen.double().filterNot { it in nonNumericDoubles }
+  private val nonMaxValueDoubles = numericDoubles.filter { it != MAX_VALUE }
+  
+  private val delta = 0.0001
+  
+  init {
+    "Exactly Matcher" - {
+      
+      "Every numeric Double" - {
+        
+        "Should be exactly" - {
+          
+          "Itself" {
+            assertAll(numericDoubles) {
+              it shouldExactlyMatch it
             }
-
-            val notAcceptedValues = table(
-                    headers("a"),
-                    row(0.1),
-                    row(1.0),
-                    row(Double.MAX_VALUE),
-                    row(Double.MIN_VALUE),
-                    row(Double.POSITIVE_INFINITY)
-            )
-
-            forNone(notAcceptedValues) {
-                it.shouldNotBePositive()
-                it shouldNotBe positive()
+          }
+        }
+        
+        
+        "Should not be exactly" - {
+          
+          "Any number smaller than itself" {
+            assertAll(nonMaxValueDoubles) {
+              it shouldNotMatchExactly it - delta
```

Comment:
Tests like this will only pass for values with an exponent relatively close to 1. For example:
`1e300 + 0.0001 === 1e300`
`1e300 + 0.0001 === 0.0001`

Reply:
I didn't quite understand what you meant.

Do you believe this test may sometimes fail?

Reply:
It looks like you're trying to create a number `it - 0.0001` that is different from `it`. But for most `Double` values, `it - 0.0001` is exactly equal to `it`.

[Double precision floating point](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) can only represent about 16 significant decimal digits. `1e300 - 0.0001` would require 300 significant decimal digits, so the result just rounds back to the original `1e300`.

So your delta needs to be relative to the value you're trying to change. For example, rather than `it - delta`, you could do something like `it * .9`, which should work for all numbers with a magnitude larger than ~1e-323 (the smallest positive denormal number).

Reply:
That's very nice to know! Learn something every day, I guess.

I'll add these corner case numbers to the generated double lists